tests/Feature/Connect/UploadAchievementTest.php

```diff
@@ -0,0 +1,895 @@
+
+        // ====================================================
+        // publish achievement
+        $params['a'] = $achievement2->ID;
+        $params['f'] = 3; // Official - hardcode for test to prevent false success if enum changes
+        $this->get($this->apiUrl('uploadachievement', $params))
+            ->assertExactJson([
+                'Success' => true,
+                'AchievementID' => $achievement2->ID,
+                'Error' => '',
+            ]);
+
+        $achievement2->refresh();
+        $this->assertEquals($achievement2->GameID, $game->ID);
+        $this->assertEquals($achievement2->Title, 'Title1');
+        $this->assertEquals($achievement2->MemAddr, '0xH0000=1');
+        $this->assertEquals($achievement2->Points, 5);
+        $this->assertEquals($achievement2->Flags, AchievementFlag::OfficialCore);
+        $this->assertNull($achievement2->type);
+        $this->assertEquals($achievement2->Author, $author->User);
+        $this->assertEquals($achievement2->BadgeName, '001234');
+
+        $game->refresh();
+        $this->assertEquals($game->achievements_published, 1);
+        $this->assertEquals($game->achievements_unpublished, 0);
+        $this->assertEquals($game->points_total, 5);
+
+        // ====================================================
+        // modify achievement
+        $params['n'] = 'Title2';
+        $params['d'] = 'Description2';
+        $params['z'] = 10;
+        $params['m'] = '0xH0001=1';
+        $params['b'] = '002345';
+        $params['x'] = 'progression';
+        $this->get($this->apiUrl('uploadachievement', $params))
+            ->assertExactJson([
+                'Success' => true,
+                'AchievementID' => $achievement2->ID,
+                'Error' => '',
+            ]);
+
+        $achievement2->refresh();
+        $this->assertEquals($achievement2->GameID, $game->ID);
+        $this->assertEquals($achievement2->Title, 'Title2');
+        $this->assertEquals($achievement2->MemAddr, '0xH0001=1');
+        $this->assertEquals($achievement2->Points, 10);
+        $this->assertEquals($achievement2->Flags, AchievementFlag::OfficialCore);
+        $this->assertEquals($achievement2->type, 'progression');
+        $this->assertEquals($achievement2->Author, $author->User);
+        $this->assertEquals($achievement2->BadgeName, '002345');
+
+        $game->refresh();
+        $this->assertEquals($game->achievements_published, 1);
+        $this->assertEquals($game->achievements_unpublished, 0);
+        $this->assertEquals($game->points_total, 10);
+
+        // ====================================================
+        // unlock achievement; contrib yield changes
+        $this->addHardcoreUnlock($author, $achievement2);
+        $this->addHardcoreUnlock($this->user, $achievement2);
+
+        $author->refresh();
+        $this->assertEquals($author->ContribCount, 1);
+        $this->assertEquals($author->ContribYield, 10);
```

Comment:
I'd skip contrib recalculation for now until we have a scalable solution for those in place.
https://github.com/RetroAchievements/RAWeb/pull/1874#discussion_r1347230085